File: csharp/DesiArchitect.DesignPatterns/07_Facade/After/BookingFacade.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Facade.After;

public class BookingFacade
{
    private readonly ISeatService _seats;
    private readonly IPaymentService _payment;
    private readonly IPnrService _pnr;
    private readonly INotificationService _notify;

    public BookingFacade(ISeatService seats, IPaymentService payment,
                         IPnrService pnr, INotificationService notify)
        => (_seats, _payment, _pnr, _notify) = (seats, payment, pnr, notify);

    public BookingResult BookTicket(BookingRequest req)
    {
        if (!_seats.Reserve(req)) return BookingResult.NoSeats;
        _payment.Charge(req);
        var pnr = _pnr.Generate(req);
        _notify.Send(req.Phone, pnr);
        return BookingResult.Confirmed(pnr);
    }
}

File: typescript/src/07-facade/after/booking-facade.ts
import type { BookingRequest, BookingResult } from '../../shared/models.js';
import { BookingResults } from '../../shared/models.js';
import type { ISeatService } from './i-seat-service.js';
import type { IPaymentService } from './i-payment-service.js';
import type { IPnrService } from './i-pnr-service.js';
import type { INotificationService } from './i-notification-service.js';

export class BookingFacade {
  constructor(
    private readonly seats: ISeatService,
    private readonly payment: IPaymentService,
    private readonly pnr: IPnrService,
    private readonly notify: INotificationService,
  ) {}

  bookTicket(req: BookingRequest): BookingResult {
    if (!this.seats.reserve(req)) return BookingResults.noSeats();
    this.payment.charge(req);
    const pnr = this.pnr.generate(req);
    this.notify.send(req.phone, pnr);
    return BookingResults.confirmed(pnr);
  }
}